fix: fix proposer boost reorg by suppressing fcu call during import

ensi321 opened the pull request:
We need to suppress fcu call during gossip block import when the incoming block is weak. This is because once notified EL, it will treat the new block as head and will not reorg back to parent if we try to reorg.

- Add `shouldOverrideForkChoiceUpdate`
- Call `shouldOverrideForkChoiceUpdate` during gossip block import, and skip fcu call if result is true
- Modify `predictProposerHead` to be a simple wrapper around `shouldOverrideForkChoiceUpdate`


Closes #7235

cc. @twoeths @nflaig

nflaig commented on packages/beacon-node/src/chain/chain.ts:
```diff
@@ -832,18 +832,12 @@ export class BeaconChain implements IBeaconChain {
-    } finally {
-      timer?.();
+  shouldOverrideForkchoiceUpdate(block: ProtoBlock): boolean {
+    const maybeOverride = this.forkChoice.shouldOverrideForkchoiceUpdate(block);
```
variable name seems strange, if this is `true` doesn't it mean we should override, "maybe override" suggests even if it is `true` we may not override

ensi321 replied:
> variable name seems strange, if this is `true` doesn't it mean we should override, "maybe override" suggests even if it is `true` we may not override

If it is true, we may or may not override (reorg). If false, it is a hard no to override (absolutely no reorg). So I think `maybeOverride` is appropriate.

nflaig replied:
but if this returns true shouldn't we always override fcU because head is weak?

ensi321 committed:
Initial commit

nflaig commented on packages/fork-choice/src/forkChoice/forkChoice.ts:
```diff
@@ -248,28 +288,24 @@ export class ForkChoice implements IForkChoice {
-    if (!currentTimeOk) {
-      return headBlock;
+    if (result.shouldOverrideFcu) {
+      this.logger?.info("Current head is weak. Predicting next block to be built on parent of head");
```
we should include block roots in the log via context

nflaig commented on packages/fork-choice/src/forkChoice/forkChoice.ts:
```diff
@@ -223,6 +224,46 @@ export class ForkChoice implements IForkChoice {
+      return {shouldOverrideFcu: false};
+    }
+
+    this.logger?.info(`Block ${blockRoot} is weak. Should override forkchoice update`);
```
would be better to use context to pass variables instead

nflaig commented on packages/beacon-node/src/chain/blocks/importBlock.ts:
```diff
@@ -212,6 +213,28 @@ export async function importBlock(
+  }
+
+  if (shouldOverrideFcu) {
+    this.logger.info("Weak block detected. Skip fcu call in importBlock");
```
should include block root(s) in log, maybe even the slot so it will show up if you filter by slot

twoeths replied:
this is more like technical information, maybe change to `verbose` level? `info` log is for users

twoeths replied:
also we wanna add a `gauge` metric in this case, that'll help us investigate later

nflaig commented on packages/fork-choice/src/forkChoice/forkChoice.ts:
```diff
@@ -248,28 +288,24 @@ export class ForkChoice implements IForkChoice {
-      return headBlock;
-    }
+    // TODO: Maybe add a transcient flag in ProtoBlock to flag `shouldOverrideFcu` block when importing
+    // so we don't need to call shouldOverrideForkChoiceUpdate again.
```
but don't we need to re-evaluate this each time? what's the current scenario in which we don't wanna call `shouldOverrideForkChoiceUpdate` again?

ensi321 replied:
~~I mean, when we import a gossip block, we have already called `shouldOverrideForkChoiceUpdate` to determine if we want to suppress fcu call. But we also call `shouldOverrideForkChoiceUpdate` during `prepareNextSlot` which is redundant.~~

~~`shouldOverrideForkChoiceUpdate` returns the same result on the same block whether it is during the first 4s, or at 8s.~~

nflaig replied:
> shouldOverrideForkChoiceUpdate returns the same result on the same block whether it is during the first 4s, or at 8s.

don't we evaluate single attestations already in fork choice which could change the result?

ensi321 replied:
> > shouldOverrideForkChoiceUpdate returns the same result on the same block whether it is during the first 4s, or at 8s.
> 
> don't we evaluate single attestations already in fork choice which could change the result?

Oh yea you are right. Forgot about single attestations

nflaig commented on packages/beacon-node/src/chain/blocks/importBlock.ts:
```diff
@@ -212,6 +213,28 @@ export async function importBlock(
   const currFinalizedEpoch = this.forkChoice.getFinalizedCheckpoint().epoch;
 
+  // Suppress fcu call if this is true
+  let shouldOverrideFcu = false;
```
why do we need to call this up here? this should be post bellatrix only

ensi321 replied:
> why do we need to call this up here? this should be post bellatrix only

Can you elaborate? `shouldOverrideFcu` will always be false pre-bellatrix, which will follow the original flow (ie. no proposer reorg)

nflaig replied:
I am mostly wondering why this logic is up here and not around L356

nflaig commented on packages/beacon-node/src/chain/blocks/importBlock.ts:
```diff
@@ -212,6 +213,28 @@ export async function importBlock(
+    this.logger.info("Weak block detected. Skip fcu call in importBlock");
+  }
+
+  // TODO: Skip this too when shouldOverrideFcu == true?
```
I don't think so cc @twoeths

twoeths replied:
yeah it's still a head so we'll still emit it
just keep it as is and drop this `TODO` comment

twoeths commented on packages/beacon-node/src/chain/blocks/importBlock.ts:
```diff
@@ -212,6 +213,28 @@ export async function importBlock(
+  let shouldOverrideFcu = false;
+
+  if (opts.isGossipBlock && isExecutionStateType(postState)) {
+    const proposerIndex = postState.epochCtx.getBeaconProposer(blockSlot + 1);
```
this `getBeaconProposer` can throw so either need to do a try/catch here
if `blockSlot` is end of this epoch we can't get proposer of next slot so maybe handle it separately

ensi321 replied:
If `blockSlot` is end of epoch, we don't reorg anyway due to this condition
https://github.com/ChainSafe/lodestar/blob/3c7d6d16870d79a8eb69b718930c171d11587f14/packages/fork-choice/src/forkChoice/forkChoice.ts#L1480

And we don't record `NotReorgedReason` here, so there is no need to handle this separately.

twoeths commented on packages/fork-choice/src/forkChoice/forkChoice.ts:
```diff
@@ -223,6 +224,46 @@ export class ForkChoice implements IForkChoice {
+  // Return true if the given block passes all criteria to be re-orged out
+  // Return false otherwise.
+  // Note when proposer boost reorg is disabled, it always returns false
+  shouldOverrideForkChoiceUpdate(currentSlot: Slot, blockRoot: RootHex): ShouldOverrideForkChoiceUpdateResult {
```
can you add a unit test for this function? also I'm curious if this function scanned through spec tests or not

ensi321 replied:
> also I'm curious if this function scanned through spec tests or not

@twoeths Hmmmm, there is, but we pass these tests even before we implemented proposer boost reorg

https://github.com/ethereum/consensus-specs/blob/6507cd11c2d5bf1a50c3de7b21555730d49b2f96/tests/core/pyspec/eth2spec/test/bellatrix/fork_choice/test_should_override_forkchoice_update.py#L36

ensi321 replied:
Unit test added.

ensi321 committed:
address comments from @nflaig

ensi321 committed:
Address @twoeths's comment

ensi321 committed:
Address @twoeths's comment

ensi321 committed:
Merge branch 'unstable' into nc/proposer-boost-reorg-fix-1

nflaig commented on packages/beacon-node/src/metrics/metrics/lodestar.ts:
```diff
@@ -741,6 +742,11 @@ export function createLodestarMetrics(
         help: "Total number of imported blobs by source",
         labelNames: ["blobsSource"],
       }),
+      notOverrideFcuReason: register.gauge<{reason: NotReorgedReason}>({
```
we should start using counter where applicable
```suggestion
      notOverrideFcuReason: register.counter<{reason: NotReorgedReason}>({
```

ensi321 committed:
Merge branch 'unstable' into nc/proposer-boost-reorg-fix-1

ensi321 committed:
Update packages/beacon-node/src/metrics/metrics/lodestar.ts

Co-authored-by: Nico Flaig <nflaig@protonmail.com>

ensi321 committed:
Move importBlock code to appropriate place

ensi321 committed:
Remove todo comment

twoeths commented on packages/beacon-node/src/chain/blocks/importBlock.ts:
```diff
@@ -225,6 +233,47 @@ export async function importBlock(
+    if (this.metrics) {
+      this.metrics.importBlock.notOverrideFcuReason.inc({reason: shouldOverrideFcuReason});
+    } else {
+      this.logger.verbose("Weak block detected. Skip fcu call in importBlock", {
```
let's have both logger and metrics, this should be very rare anyway

twoeths commented on packages/beacon-node/src/chain/blocks/importBlock.ts:
```diff
@@ -225,6 +233,47 @@ export async function importBlock(
   const newHead = this.recomputeForkChoiceHead(ForkchoiceCaller.importBlock);
   const currFinalizedEpoch = this.forkChoice.getFinalizedCheckpoint().epoch;
 
+  // Suppress fcu call if this is true
```
add context of this issue to this comment, also refer to the issue itself to make it easier to reason about later

twoeths commented on packages/beacon-node/src/chain/blocks/importBlock.ts:
```diff
@@ -225,6 +233,47 @@ export async function importBlock(
+
+  if (shouldOverrideFcu) {
+    if (this.metrics) {
+      this.metrics.importBlock.notOverrideFcuReason.inc({reason: shouldOverrideFcuReason});
```
the name is inconsistent: `notOverrideFcuReason` vs `shouldOverrideFcuReason`

ensi321 committed:
Add unit test

ensi321 committed:
Address @twoeths comment

nflaig committed:
Merge branch 'unstable' into nc/proposer-boost-reorg-fix-1

twoeths commented on packages/fork-choice/src/forkChoice/forkChoice.ts:
```diff
@@ -223,6 +224,50 @@ export class ForkChoice implements IForkChoice {
+    }
+    // Skip re-org attempt if proposer boost (reorg) are disabled
+    if (!this.opts?.proposerBoost || !this.opts?.proposerBoostReorg) {
+      this.logger?.verbose("Skip shouldOverrideForkChoiceUpdate check since the related flags are disabled");
```
should log both block slot + root, check other log statements as well

twoeths commented on packages/beacon-node/src/chain/blocks/importBlock.ts:
```diff
@@ -314,9 +322,52 @@ export async function importBlock(
+      this.metrics.importBlock.notOverrideFcuReason.inc({reason: notOverrideFcuReason});
+    }
+    this.logger.verbose("Weak block detected. Skip fcu call in importBlock", {
+      blockRoot: blockRootHex,
```
need to also log slot

twoeths reviewed (commented):
I'd like to have both block root + slot in all log statements so that we can debug easier later, looks good to me other than that

ensi321 committed:
Log slot

nflaig commented on packages/beacon-node/src/chain/blocks/importBlock.ts:
```diff
@@ -314,9 +322,53 @@ export async function importBlock(
+  if (shouldOverrideFcu) {
+    if (this.metrics) {
+      this.metrics.importBlock.notOverrideFcuReason.inc({reason: notOverrideFcuReason});
+    }
```
```suggestion
    this.metrics?.importBlock.notOverrideFcuReason.inc({reason: notOverrideFcuReason});
```

nflaig commented on packages/beacon-node/src/chain/blocks/importBlock.ts:
```diff
@@ -314,9 +322,53 @@ export async function importBlock(
+      if (isStartSlotOfEpoch(blockSlot + 1)) {
+        notOverrideFcuReason = NotReorgedReason.NotShufflingStable;
+      } else {
+        this.logger.warn("Unable to get beacon proposer. Do not override fcu.", {slot: blockSlot + 1}, e as Error);
```
`slot` here might be misleading as it's the proposal slot, not block slot or clock slot

maybe add const `proposalSlot` as we do `blockSlot + 1` in multiple places
```suggestion
        this.logger.warn("Unable to get beacon proposer. Do not override fcu.", {proposalSlot}, e as Error);
```

nflaig commented on packages/beacon-node/src/chain/blocks/importBlock.ts:
```diff
@@ -314,9 +322,53 @@ export async function importBlock(
+
+  if (shouldOverrideFcu) {
+    if (this.metrics) {
+      this.metrics.importBlock.notOverrideFcuReason.inc({reason: notOverrideFcuReason});
```
why would we set `notOverrideFcuReason` (reason why we are not overriding) if `shouldOverrideFcu == true` meaning we override fcu call?

(same for the [log below](https://github.com/ChainSafe/lodestar/blob/692eaacd20e905da312b234cb315aec63536fa23/packages/beacon-node/src/chain/blocks/importBlock.ts#L361-L365))

nflaig commented on packages/fork-choice/src/forkChoice/forkChoice.ts:
```diff
@@ -223,6 +224,52 @@ export class ForkChoice implements IForkChoice {
+    if (!this.opts?.proposerBoost || !this.opts?.proposerBoostReorg) {
+      this.logger?.verbose("Skip shouldOverrideForkChoiceUpdate check since the related flags are disabled", {
+        slot: currentSlot,
+      });
```
could print out flags here

```suggestion
      this.logger?.verbose("Skip shouldOverrideForkChoiceUpdate check since the related flags are disabled", {
        slot: currentSlot,
        proposerBoost: this.opts?.proposerBoost,
        proposerBoostReorg: this.opts?.proposerBoostReorg,
      });
```

could destructure opts above if-statement

nflaig commented on packages/fork-choice/src/forkChoice/forkChoice.ts:
```diff
@@ -223,6 +224,52 @@ export class ForkChoice implements IForkChoice {
   }
 
+  // Basically called by `predictProposerHead`. If the result is not same as blockRoot's block, return true else false
+  // See https://github.com/ethereum/consensus-specs/blob/a94d85bb3b33f84a5830eb98945b790b02ae80e8/specs/bellatrix/fork-choice.md#should_override_forkchoice_update
```
```suggestion
  // See https://github.com/ethereum/consensus-specs/blob/v1.5.0/specs/bellatrix/fork-choice.md#should_override_forkchoice_update
```

nflaig commented on packages/beacon-node/src/metrics/metrics/lodestar.ts:
```diff
@@ -744,6 +745,11 @@ export function createLodestarMetrics(
         labelNames: ["blobsSource"],
       }),
+      notOverrideFcuReason: register.counter<{reason: NotReorgedReason}>({
+        name: "beacon_import_block_not_override_fcu_reason_total",
```
```suggestion
        name: "lodestar_import_block_not_override_fcu_reason_total",
```

nflaig commented on packages/fork-choice/src/forkChoice/forkChoice.ts:
```diff
@@ -248,28 +294,26 @@ export class ForkChoice implements IForkChoice {
     }
 
-    const {prelimProposerHead} = this.getPreliminaryProposerHead(headBlock, parentBlock, proposalSlot);
+    currentSlot = currentSlot ?? this.fcStore.currentSlot;
```
why is passing `currentSlot` optional? doesn't look like we call this function without passing it

nflaig commented on packages/fork-choice/src/forkChoice/forkChoice.ts:
```diff
@@ -223,6 +224,52 @@ export class ForkChoice implements IForkChoice {
     }
   }
 
+  // Basically called by `predictProposerHead`. If the result is not same as blockRoot's block, return true else false
```
we are calling this separately now as well, this kinda contradicts with the other comment here
https://github.com/ChainSafe/lodestar/blob/692eaacd20e905da312b234cb315aec63536fa23/packages/fork-choice/src/forkChoice/interface.ts#L117-L120

nflaig commented on packages/fork-choice/test/unit/forkChoice/shouldOverrideForkChoiceUpdate.test.ts:
```diff
@@ -0,0 +1,220 @@
+      expectReorg: true,
+    },
+    {
+      id: "No reorg when head block is timly",
```
```suggestion
      id: "No reorg when head block is timely",
```

nflaig commented on packages/fork-choice/test/unit/forkChoice/shouldOverrideForkChoiceUpdate.test.ts:
```diff
@@ -0,0 +1,220 @@
+    currentSlot: blockSeenSlot,
+    expectedNotReorgedReason,
+  } of testCases) {
+    it(`${id}`, async () => {
```
```suggestion
    it(id, async () => {
```

nflaig commented on packages/fork-choice/src/forkChoice/forkChoice.ts:
```diff
@@ -223,6 +224,52 @@ export class ForkChoice implements IForkChoice {
+  shouldOverrideForkChoiceUpdate(currentSlot: Slot, blockRoot: RootHex): ShouldOverrideForkChoiceUpdateResult {
+    const headBlock = this.getBlockHex(blockRoot);
+    if (headBlock === null) {
+      // should not happen beacause this block just got imported. Fall back to no-reorg.
```
```suggestion
      // should not happen because this block just got imported. Fall back to no-reorg.
```

nflaig committed:
Merge branch 'unstable' into nc/proposer-boost-reorg-fix-1

nflaig committed:
chore: add `should_override_forkchoice_update` spec test check (#8020)

**Motivation**

We are not running `should_override_forkchoice_update` spec tests from
https://github.com/ethereum/consensus-specs/pull/3034 as checks are
missing.

**Description**

- add `should_override_forkchoice_update` spec test check
- pass `secFromSlot` to `shouldOverrideForkChoiceUpdate` as it's
required now
- add [`proposal_slot == current_slot and
proposing_on_time`](https://github.com/ethereum/consensus-specs/blob/18387696969c0bb34e96164434a3a36edca296c9/specs/bellatrix/fork-choice.md?plain=1#L145)
to `shouldOverrideForkChoiceUpdate` to pass spec tests

ensi321 committed:
Update packages/beacon-node/src/chain/blocks/importBlock.ts

Co-authored-by: Nico Flaig <nflaig@protonmail.com>

ensi321 committed:
Update packages/beacon-node/src/chain/blocks/importBlock.ts

Co-authored-by: Nico Flaig <nflaig@protonmail.com>

ensi321 committed:
Update packages/fork-choice/src/forkChoice/forkChoice.ts

Co-authored-by: Nico Flaig <nflaig@protonmail.com>

ensi321 committed:
Update packages/fork-choice/src/forkChoice/forkChoice.ts

Co-authored-by: Nico Flaig <nflaig@protonmail.com>

ensi321 committed:
Apply suggestions from code review

Co-authored-by: Nico Flaig <nflaig@protonmail.com>

nflaig committed:
Add var for proposalSlot

nflaig committed:
Update flags disabled check

ensi321 committed:
Address @nflaig's comment

philknows commented:
Would like an update on this for #7926 and its testing status on `feat4`

nflaig committed:
Remove commented out test case

nflaig committed:
Fix debug log in spec test

nflaig committed:
Merge branch 'unstable' into nc/proposer-boost-reorg-fix-1

nflaig reviewed (dismissed):
LGTM - we should consider enabling proposer boost reorg by default

nflaig committed:
Update not override fcu reason if we are not proposer of next slot

wemeetagain commented:
:tada: This PR is included in [v1.33.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.33.0) :tada:
